feedback/ngffeedback.cpp

```diff
@@ -0,0 +1,326 @@
+            this, &NGFFeedback::playing);
+    connect(&m_client, &Ngf::Client::eventPaused,
+            this, &NGFFeedback::paused);
+
+    m_effects[QFeedbackEffect::Press] = QStringLiteral("feedback_press");
+    m_effects[QFeedbackEffect::Release] = QStringLiteral("feedback_release");
+    m_effects[QFeedbackEffect::PressWeak] = QStringLiteral("feedback_press_weak");
+    m_effects[QFeedbackEffect::ReleaseWeak] = QStringLiteral("feedback_release_weak");
+    m_effects[QFeedbackEffect::PressStrong] = QStringLiteral("feedback_press_strong");
+    m_effects[QFeedbackEffect::ReleaseStrong] = QStringLiteral("feedback_release_strong");
+    m_effects[QFeedbackEffect::DragStart] = QStringLiteral("feedback_drag_start");
+    m_effects[QFeedbackEffect::DragDropInZone] = QStringLiteral("feedback_drag_drop_in_zone");
+    m_effects[QFeedbackEffect::DragDropOutOfZone] = QStringLiteral("feedback_drag_drop_out_of_zone");
+    m_effects[QFeedbackEffect::DragCrossBoundary] = QStringLiteral("feedback_drag_cross_boundary");
+    m_effects[QFeedbackEffect::Appear] = QString();
+    m_effects[QFeedbackEffect::Disappear] = QString();
+    m_effects[QFeedbackEffect::Move] = QString();
+}
+
+NGFFeedback::~NGFFeedback()
+{
+    qCDebug(ngflc) << "Deinitializing plugin";
+}
+
+void NGFFeedback::failed(quint32 id)
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCWarning(ngflc) << "Effect failed, id" << id;
+        m_activeEffects.removeAll(*active);
+    }
+}
+
+void NGFFeedback::completed(quint32 id)
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCDebug(ngflc) << "Effect completed, id" << id;
+        m_activeEffects.removeAll(*active);
+    }
+}
+
+void NGFFeedback::playing(quint32 id)
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCDebug(ngflc) << "Effect playing, id" << id;
+        active->state = QFeedbackEffect::Running;
+    }
+}
+
+void NGFFeedback::paused(quint32 id)
+{
+    ActiveEffect *active = findEffect(id);
+    if (active) {
+        qCDebug(ngflc) << "Effect paused, id" << id;
+        active->state = QFeedbackEffect::Paused;
+    }
+}
+
+bool NGFFeedback::play(QFeedbackEffect::Effect effect)
+{
+    quint32 id;
+    switch (effect) {
+    case QFeedbackEffect::Press:
+    case QFeedbackEffect::Release:
+    case QFeedbackEffect::PressWeak:
+    case QFeedbackEffect::ReleaseWeak:
+    case QFeedbackEffect::PressStrong:
+    case QFeedbackEffect::ReleaseStrong:
+    case QFeedbackEffect::DragStart:
+    case QFeedbackEffect::DragDropInZone:
+    case QFeedbackEffect::DragDropOutOfZone:
+    case QFeedbackEffect::DragCrossBoundary:
+        qCDebug(ngflc) << "Playing effect #" << effect << "(" << m_effects[effect] << ")";
+        id = m_client.play(m_effects[effect]);
+        if (!id)
+            qCWarning(ngflc) << "Could not play effect";
+        else
+            m_activeEffects.append({ id, QFeedbackEffect::Running, nullptr });
```

Comment:
Yes, I understood that and it probably doesn't serve any purpose to store the status (since you can't actually retrieve it), but logging the events stopping might be useful still. If we don't care about logging that, then I can just drop this branch.

Reply:
Maybe log the id of started effect, then those can be matched by logging the id on effect states, even without having them in the array.

Reply:
Right, I changed this and the signal handlers. I also added a comment about this.